dashboard/app/models/school_district.rb

```diff
@@ -89,25 +98,101 @@ def self.seed_from_s3
+  # @param filepath [String] Filepath for S3 object.
+  # @param import_options [Hash] The CSV file parsing options.
+  # @param is_dry_run [Boolean] Allows a "dry run" of seeding a CSV without making database writes.
+  # @param new_attributes  [Array] List of attributes included in a given import that are new to the model, and thus should not be used to determine whether a record is being "updated" or "unchanged"
+  # @param parse_row [Block] A block to parse a row of new data -- see School.seed_from_s3 for examples.
+  def self.seed_s3_object(bucket, filepath, import_options, is_dry_run: false, new_attributes: [], &parse_row)
+    AWS::S3.seed_from_file(bucket, filepath, is_dry_run) do |filename|
+      merge_from_csv(
+        filename,
+        import_options,
+        true,
+        is_dry_run: is_dry_run,
+        new_attributes: new_attributes,
+        &parse_row
+      )
     end
   end
 
-  # Loads/merges the data from a CSV into the schools table.
+  # Loads/merges the data from a CSV into the school districts table.
   # Requires a block to parse the row.
   # @param filename [String] The CSV file name.
   # @param options [Hash] The CSV file parsing options.
   # @param write_updates [Boolean] Specify whether existing rows should be updated.  Default to true for backwards compatible with existing logic that calls this method to UPSERT school districts.
-  def self.merge_from_csv(filename, options = CSV_IMPORT_OPTIONS, write_updates = true)
-    CSV.read(filename, options).each do |row|
-      parsed = block_given? ? yield(row) : row.to_hash.symbolize_keys
-      loaded = find_by_id(parsed[:id])
-      if loaded.nil?
-        SchoolDistrict.new(parsed).save!
-      elsif write_updates
-        loaded.assign_attributes(parsed)
-        loaded.update!(parsed) if loaded.changed?
+  # @param is_dry_run [Boolean] Specify that this is a dry run, and no writes to the database should be conducted. Gives more detailed output of expected changes from importing the given CSV.
+  # @param new_attributes  [Array] List of attributes included in a given import that are new to the model, and thus should not be used to determine whether a record is being "updated" or "unchanged"
+  def self.merge_from_csv(filename, options = CSV_IMPORT_OPTIONS, write_updates = true, is_dry_run: false, new_attributes: [])
+    districts = nil
+    new_districts = []
+    updated_districts = 0
+    unchanged_districts = 0
+
+    ActiveRecord::Base.transaction do
+      districts = CSV.read(filename, options).each do |row|
+        parsed = block_given? ? yield(row) : row.to_hash.symbolize_keys
+        loaded = find_by_id(parsed[:id])
+        if loaded.nil?
+          SchoolDistrict.new(parsed).save! unless is_dry_run
+          new_districts << parsed
+        elsif write_updates
+          loaded.assign_attributes(parsed)
+          if loaded.changed?
+            loaded.changed.sort == new_attributes.sort ?
```

Comment:
can you clarify how this line checks if the district is changed?

Reply:
If the attributes that have been changed (`loaded.changed`) are only the provided attributes that are being imported for the first time (`new_attributes`), then treat the row as "unchanged" -- in this import, there's a column that's never been imported before (`last_known_school_year_open`), so I wanted to see which rows had changes to existing columns vs. just were putting in values for last known school year open.